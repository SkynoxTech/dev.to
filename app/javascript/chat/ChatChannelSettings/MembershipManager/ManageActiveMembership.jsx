import { h, Component } from 'preact';
import PropTypes from 'prop-types';
import { defaulMembershipPropType } from '../../../common-prop-types/membership-prop-type';
import MembershipSection from './MembershipSection';

export default class ManageActiveMembership extends Component {
  static propTypes = {
    activeMemberships: PropTypes.arrayOf(defaulMembershipPropType).isRequired,
    currentMembership: PropTypes.isRequired,
    invitationLink: PropTypes.string.isRequired,
    removeMembership: PropTypes.func.isRequired,
    handleUpdateMembershipRole: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      activeMemberships: props.activeMemberships,
      searchMembers: null,
      listAllMemberShips: props.activeMemberships,
      currentMembership: props.currentMembership,
      invitationLink: props.invitationLink,
      removeMembership: props.removeMembership,
      handleUpdateMembershipRole: props.handleUpdateMembershipRole,
    };
  }

  componentWillReceiveProps() {
    const {
      activeMemberships,
      currentMembership,
      removeMembership,
      invitationLink,
    } = this.props;
    this.setState({
      listAllMemberShips: activeMemberships,
      currentMembership,
      invitationLink,
      removeMembership,
      activeMemberships,
    });
  }

  searchTheMembershipUser = (e) => {
    const query = e.target?.value?.toLowerCase();

    this.setState((prevState) => {
      const filteredActiveMemberships = prevState.activeMemberships.filter(
        (activeMembership) => {
          const value = activeMembership.name.toLowerCase();
          return value.includes(query);
        },
      );

      return {
        searchMembers: query,
        listAllMemberShips: filteredActiveMemberships,
      };
    });
  };

  render() {
    const {
      searchMembers,
      listAllMemberShips,
      currentMembership,
      removeMembership,
      handleUpdateMembershipRole,
    } = this.state;

    return (
      <div className="pt-3">
        <div className="p-4 grid gap-2 crayons-card my-4 mx-auto membership-manager">
          <h2 className="text-title">Chat Channel Membership manager</h2>
        </div>
        <div className="p-4 grid gap-2 crayons-card my-4">
          <input
            type="text"
            className="crayons-textfield"
            placeholder="Search Member..."
            value={searchMembers}
            name="search-members"
            onKeyUp={this.searchTheMembershipUser.bind(this)}
            aria-label="search memberships"
          />
          <div
            data-spy="scroll"
            data-offset="5"
            className="chat_channel-member-list"
          >
            <MembershipSection
              memberships={listAllMemberShips}
              currentMembership={currentMembership}
              removeMembership={removeMembership}
              handleUpdateMembershipRole={handleUpdateMembershipRole}
            />
          </div>
        </div>
      </div>
    );
  }
}
