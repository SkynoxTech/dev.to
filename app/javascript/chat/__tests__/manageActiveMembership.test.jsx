import { h } from 'preact';
import render from 'preact-render-to-json';
import { shallow } from 'preact-render-spy';
import ManageActiveMembership from '../ChatChannelSettings/MembershipManager/ManageActiveMembership';

const data = {
  currentMembership: {
    role: 'mod',
    user_id: 1,
    chat_channel_id: 1,
    id: 2,
    status: 'active',
  },
  activeMemberships: [
    {
      name: 'test user',
      username: 'testusername',
      user_id: '1',
      membership_id: '2',
      role: 'mod',
      status: 'active',
      image: '',
    },
  ],
  invitationLink: 'https://dummy.com',
};

const getManageActiveMembership = (resource) => (
  <ManageActiveMembership
    currentMembership={resource.currentMembership}
    activeMemberships={resource.activeMemberships}
    invitationLink={resource.invitationLink}
  />
);

describe('<ManageActiveMembership />', () => {
  it('should render & test snapshot', () => {
    const tree = render(getManageActiveMembership(data));
    expect(tree).toMatchSnapshot();
  });

  it('should render the active membership', () => {
    const context = shallow(getManageActiveMembership(data));
    expect(context.find('.membership-manager').exists()).toEqual(true);
  });

  it('should render the active membership title', () => {
    const context = shallow(getManageActiveMembership(data));
    expect(context.find('.chat_channel-member-list').exists()).toEqual(true);
  });
});
