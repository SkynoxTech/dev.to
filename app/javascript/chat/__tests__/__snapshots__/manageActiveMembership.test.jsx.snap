// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageActiveMembership /> should render & test snapshot 1`] = `
<div
  class="pt-3"
>
  <div
    class="p-4 grid gap-2 crayons-card my-4 mx-auto membership-manager"
  >
    <h2
      class="text-title"
    >
      Chat Channel Membership manager
    </h2>
  </div>
  <div
    class="p-4 grid gap-2 crayons-card my-4"
  >
    <input
      class="crayons-textfield"
      name="search-members"
      onKeyUp={[Function]}
      placeholder="Search Member..."
      type="text"
      value={null}
    />
    <div
      class="chat_channel-member-list"
      data-offset="5"
      data-spy="scroll"
    >
      <div
        class="membership-section"
      >
        <div
          class="flex items-center my-3 member-list-item justify-content-between"
        >
          <div
            class="w-100"
          >
            <a
              class="chatmessagebody__username--link"
              data-content="sidecar-user"
              href="/testusername"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="crayons-avatar crayons-avatar--l mr-3"
              >
                <img
                  alt="test user profile"
                  class="crayons-avatar__image align-middle"
                  role="presentation"
                  src=""
                />
              </span>
              <span
                class="mr-2 user_name"
              >
                test user
              </span>
              <span>
                <img
                  alt="admin emoji"
                  class="admin-emoji"
                  data-content="admin emoji"
                  src=""
                  title="MOD"
                />
              </span>
            </a>
          </div>
          <div
            class="flex-shrink-1"
          >
            <div
              class="membership-actions"
            >
              <span
                class="membership-management__dropdown-button"
              >
                <img
                  alt="channel config"
                  class="w-25"
                  data-content="drop-down-image"
                  src=""
                />
              </span>
              <div
                class="membership-management__dropdown-memu"
              >
                <button
                  class="crayons-btn crayons-btn--ghost remove-membership p-2"
                  data-membership-id="2"
                  data-role="member"
                  type="button"
                >
                  Add as Member
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="p-4 grid gap-2 crayons-card my-4 invitation-section"
  >
    <h3
      class="text-center title"
    >
      Invitation Link
    </h3>
    <clipboard-copy
      aria-controls="image-markdown-copy-link-announcer"
      aria-live="polite"
      class="flex items-center flex-1"
      for="chat-channel-unviation-url"
      onClick={[Function]}
    >
      <input
        class="crayons-textfield mr-2"
        id="chat-channel-unviation-url"
        readOnly="true"
        type="text"
        value="https://dummy.com"
      />
      <button
        class="crayons-btn crayons-btn--ghost crayons-btn--icon spec__image-markdown-copy"
        disabled={false}
        type="button"
      >
        <svg
          aria-labelledby="fc5f15add1e114844f5e"
          class="crayons-icon copy-icon"
          height="24"
          role="img"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="fc5f15add1e114844f5e"
          >
            Copy Markdown for image
          </title>
          <path
            d="M7 6V3a1 1 0 011-1h12a1 1 0 011 1v14a1 1 0 01-1 1h-3v3c0 .552-.45 1-1.007 1H4.007A1 1 0 013 21l.003-14c0-.552.45-1 1.007-1H7zm2 0h8v10h2V4H9v2zm-2 5v2h6v-2H7zm0 4v2h6v-2H7z"
          />
        </svg>
      </button>
      <span
        class="fs-s opacity-0"
        id="image-markdown-copy-link-announcer"
        role="alert"
      >
        Copied!
      </span>
    </clipboard-copy>
  </div>
</div>
`;
